flang/lib/Lower/FIRBuilder.cpp

```diff
@@ -218,3 +219,45 @@ Fortran::lower::FirOpBuilder::createShape(mlir::Location loc,
                      return mlir::Value{};
                    });
 }
+
+mlir::Value
+Fortran::lower::FirOpBuilder::createBox(mlir::Location loc,
+                                        const fir::ExtendedValue &exv) {
+  auto itemAddr = fir::getBase(exv);
+  auto elementType = fir::dyn_cast_ptrEleTy(itemAddr.getType());
+  if (!elementType)
+    mlir::emitError(loc, "internal: expected a memory reference type ")
+        << itemAddr.getType();
+  auto boxTy = fir::BoxType::get(elementType);
+  return exv.match(
+      [&](const fir::ArrayBoxValue &box) -> mlir::Value {
+        auto s = createShape(loc, exv);
+        return create<fir::EmboxOp>(loc, boxTy, itemAddr, s);
+      },
+      [&](const fir::CharArrayBoxValue &box) -> mlir::Value {
+        auto s = createShape(loc, exv);
+        if (Fortran::lower::CharacterExprHelper::hasConstantLengthInType(exv))
+          return create<fir::EmboxOp>(loc, boxTy, itemAddr, s);
+
+        mlir::Value emptySlice;
+        llvm::SmallVector<mlir::Value, 1> lenParams{box.getLen()};
+        return create<fir::EmboxOp>(loc, boxTy, itemAddr, s, emptySlice,
+                                    lenParams);
+      },
+      [&](const fir::BoxValue &box) -> mlir::Value {
+        auto s = createShape(loc, exv);
+        // FIXME: get character dynamic length if needed.
```

Comment:
This comment seems suspect. A scalar CHARACTER is encoded as a CharBoxValue and an array of CHARACTER is encoded as a CharArrayBoxValue. I think it's a bug if a CHARACTER variable is encoded otherwise.

Reply:
This will need to support type parameters and origin shifts as well.

Reply:
Ok, since we are not using it yet, it was not entirely clear to me that we would not use BoxValue for weird character, but apart for something like an assumed rank character (and coarrays ?), I think CharArrayBoxValue/CharBoxValue  should cover all other cases. I have removed the comment.